### PR TITLE
Vet connection requests and improve in-order message delivery

### PR DESCRIPTION
**Pull request by michaelsproul**
This PR includes 2 main changes on top of the `quorum-to` branch:

1. Only connect back to nodes when we receive a `Connect` message if we think we should be connected to them. This seems to fix connect-disconnect loops.
2. Deliver messages in-order, _even if they were sent at the same step_. Previously we allowed messages sent at the same step to get out of order, and as part of debugging the implementation of part (1) I figured I would fix this. I think it makes sense for a `Connect` and a `Disconnect` sent on the same step (either side of an `update_state()` call perhaps) to be delivered in order so that the node at the other end doesn't get confused...

### CODE REVIEW

File: src/node.rs [hunk 1]
```diff
@@ -729,6 +729,11 @@ impl Node {
             .collect()
     }
 
+    fn should_be_connected(&self, node: Name, blocks: &Blocks) -> bool {
+        let neighbours = nodes_in_any(blocks, &self.current_blocks);
+        neighbours.contains(&node)
```
Comment (fizyk20) [inline src/node.rs]: How do we handle newly joining nodes, which aren't yet a part of any block, but we should connect to them since they will be a part of our section? Is that somehow a different situation?

Comment (michaelsproul) [inline src/node.rs]: It's a bit dodgy, but we connect to joining nodes no matter what when we get their `NodeJoined` message... There's a `TODO` in the code to change it so that only the relevant nodes get that initial `NodeJoined` message, but I think we can leave that for another PR.

**Review by fizyk20 — APPROVED** [review]
